api/v1_projects.inc

```diff
@@ -116,37 +116,131 @@ function api_v1_projects($method, $data, $query_params)
 //---------------------------------------------------------------------------
 // projects/:projectid
 
+// Return a list of updatable project fields mapped from the API names
+// to the Project object names.
+function get_updatable_project_fields()
+{
+    return [
+        // API name => Project name
+        "title" => "nameofwork",
+        "author" => "authorsname",
+        "languages" => "languages",
+        "genre" => "genre",
+        "difficulty" => "difficulty",
+        "special_day" => "special_code",
+        "project_manager" => "username",
+        "image_source" => "image_source",
+        "image_preparer" => "image_preparer",
+        "text_preparer" => "text_preparer",
+        "pg_ebook_number" => "postednum",
+        "comments" => "comments",
+        "comment_format" => "comment_format",
+        "clearance" => "clearance",
+        "custom_characters" => "custom_chars",
+    ];
+}
+
+function create_or_update_project($project)
+{
+    // can the user manage the existing project?
+    if ($project->projectid && !$project->can_be_managed_by_current_user) {
+        throw new UnauthorizedError();
+    }
+
+    // can the user create new projects?
+    if (!$project->projectid && user_has_project_loads_disabled()) {
+        throw new UnauthorizedError();
+    }
+
+    // update all updatable fields
+    $updates = (array)api_get_request_body();
+    try {
+        foreach (get_updatable_project_fields() as $api_key => $project_key) {
+            // allow partial record updates by only updating specified values
+            if (isset($updates[$api_key])) {
+                $project->$project_key = $updates[$api_key];
+            }
+        }
+
+        $project->validate(true);
+    } catch (ValueError $error) {
+        throw new InvalidValue($error->getMessage());
+    }
+
+    // validate charsuites
+    if (!$project->projectid || isset($updates["character_suites"])) {
+        if (!is_array($updates["character_suites"])) {
+            throw new InvalidValue("character_suites must be an array.");
+        }
+
+        if (sizeof($updates["character_suites"]) == 0) {
+            throw new InvalidValue("At least one Character Suite is required.");
+        }
+
+        $invalid_charsuites = array_diff($updates["character_suites"],
+            array_extract_field(CharSuites::get_all(), "name"));
+        if ($invalid_charsuites) {
+            throw new InvalidValue(sprintf("%s is/are not valid Character Suites.", join(", ", $invalid_charsuites)));
+        }
+    }
+
+    $project->save();
+
+    if (isset($updates["character_suites"])) {
+        $project->set_charsuites($updates["character_suites"]);
```

Comment:
This is an oddity in how character suites are associated with projects. Character suites are linked to projects via a linking table, not stored in the `project` table directly. Because of this, we can't set a project's character suites until a project has been created via `save()`. This also means that `Project->validate()` doesn't validate character suites because it is only interested in values for the projects table.

I considered splitting out character suites into a separate endpoint, but projects shouldn't be created without a character suite.

This seemed like the least bad solution to this mess. I'm open to suggestions!